File: packages/Autonomo.CSharp/Instance.cs
namespace Autonomo;

/// <summary>
/// Platform type for the instance
/// </summary>
public enum Platform
{
    Web,
    Mobile,
    Desktop
}

/// <summary>
/// Configuration for initializing an app instance
/// </summary>
public class InstanceConfig
{
    public string Name { get; init; } = "";
    public Platform Platform { get; init; }
    public string? InstanceId { get; init; }
    public string? Version { get; init; }
    public Dictionary<string, object>? Meta { get; init; }
}

/// <summary>
/// Information about this app instance
/// </summary>
public class InstanceInfo
{
    public string InstanceId { get; init; } = "";
    public string Name { get; init; } = "";
    public string BridgeId { get; init; } = "";
    public Platform Platform { get; init; }
    public string? Version { get; init; }
    public long CreatedAt { get; init; }
    public Dictionary<string, object>? Meta { get; init; }

    public Dictionary<string, object> ToDictionary()
    {
        var result = new Dictionary<string, object>
        {
            ["instanceId"] = InstanceId,
            ["name"] = Name,
            ["bridgeId"] = BridgeId,
            ["platform"] = Platform.ToString().ToLower(),
            ["createdAt"] = CreatedAt
        };
        if (Version != null) result["version"] = Version;
        if (Meta != null) result["meta"] = Meta;
        return result;
    }
}

/// <summary>
/// Singleton instance manager
/// </summary>
public sealed class InstanceManager
{
    private static readonly Lazy<InstanceManager> _instance = new(() => new InstanceManager());
    public static InstanceManager Instance => _instance.Value;

    private InstanceInfo? _current;
    private readonly object _lock = new();

    private InstanceManager() { }

    /// <summary>
    /// Generate a short unique ID
    /// </summary>
    private static string GenerateInstanceId()
    {
        return Guid.NewGuid().ToString("N")[..8];
    }

    /// <summary>
    /// Initialize this app instance.
    /// 
    /// Call once at app startup. Each process gets a unique instance ID.
    /// </summary>
    /// <example>
    /// InstanceManager.Instance.InitInstance(new InstanceConfig
    /// {
    ///     Name = "my-app",
    ///     Platform = Platform.Desktop
    /// });
    /// </example>
    public InstanceInfo InitInstance(InstanceConfig config)
    {
        lock (_lock)
        {
            var instanceId = config.InstanceId ?? GenerateInstanceId();
            _current = new InstanceInfo
            {
                InstanceId = instanceId,
                Name = config.Name,
                BridgeId = $"{config.Name}-{instanceId}",
                Platform = config.Platform,
                Version = config.Version,
                CreatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                Meta = config.Meta
            };
            Console.WriteLine($"[Autonomo] Instance initialized: {_current.BridgeId}");
            return _current;
        }
    }

    /// <summary>
    /// Get the current instance info
    /// </summary>
    public InstanceInfo? GetInstance()
    {
        lock (_lock)
        {
            return _current;
        }
    }

    /// <summary>
    /// Get the current instance info or throw
    /// </summary>
    public InstanceInfo RequireInstance()
    {
        lock (_lock)
        {
            if (_current == null)
            {
                throw new InvalidOperationException("Autonomo instance not initialized. Call InitInstance() first.");
            }
            return _current;
        }
    }

    /// <summary>
    /// Get just the bridge ID
    /// </summary>
    public string? GetBridgeId()
    {
        lock (_lock)
        {
            return _current?.BridgeId;
        }
    }

    /// <summary>
    /// Reset the instance (mainly for testing)
    /// </summary>
    public void ResetInstance()
    {
        lock (_lock)
        {
            _current = null;
        }
    }
}

File: packages/autonomo-kotlin/src/main/kotlin/com/autonomo/Instance.kt
package com.autonomo

import java.util.UUID

/**
 * Platform type for the instance
 */
enum class Platform(val value: String) {
    WEB("web"),
    MOBILE("mobile"),
    DESKTOP("desktop")
}

/**
 * Configuration for initializing an app instance
 */
data class InstanceConfig(
    val name: String,
    val platform: Platform,
    val instanceId: String? = null,
    val version: String? = null,
    val meta: Map<String, Any>? = null
)

/**
 * Information about this app instance
 */
data class InstanceInfo(
    val instanceId: String,
    val name: String,
    val bridgeId: String,
    val platform: Platform,
    val version: String? = null,
    val createdAt: Long = System.currentTimeMillis(),
    val meta: Map<String, Any>? = null
) {
    fun toMap(): Map<String, Any?> = buildMap {
        put("instanceId", instanceId)
        put("name", name)
        put("bridgeId", bridgeId)
        put("platform", platform.value)
        put("createdAt", createdAt)
        version?.let { put("version", it) }
        meta?.let { put("meta", it) }
    }
}

/**
 * Singleton instance manager
 *
 * Manages the unique identity of this app instance.
 */
object InstanceManager {
    @Volatile
    private var currentInstance: InstanceInfo? = null
    private val lock = Any()

    /**
     * Generate a short unique ID
     */
    private fun generateInstanceId(): String {
        return UUID.randomUUID().toString().take(8)
    }

    /**
     * Initialize this app instance.
     *
     * Call once at app startup. Each process gets a unique instance ID.
     *
     * ```kotlin
     * // In Application.onCreate() or Activity.onCreate()
     * InstanceManager.initInstance(InstanceConfig(
     *     name = "my-app",
     *     platform = Platform.MOBILE
     * ))
     * ```
     */
    fun initInstance(config: InstanceConfig): InstanceInfo {
        synchronized(lock) {
            val instanceId = config.instanceId ?: generateInstanceId()
            val info = InstanceInfo(
                instanceId = instanceId,
                name = config.name,
                bridgeId = "${config.name}-$instanceId",
                platform = config.platform,
                version = config.version,
                meta = config.meta
            )
            currentInstance = info
            println("[Autonomo] Instance initialized: ${info.bridgeId}")
            return info
        }
    }

    /**
     * Get the current instance info
     */
    fun getInstance(): InstanceInfo? {
        synchronized(lock) {
            return currentInstance
        }
    }

    /**
     * Get the current instance info or throw
     */
    fun requireInstance(): InstanceInfo {
        synchronized(lock) {
            return currentInstance
                ?: throw IllegalStateException("Autonomo instance not initialized. Call initInstance() first.")
        }
    }

    /**
     * Get just the bridge ID
     */
    fun getBridgeId(): String? {
        synchronized(lock) {
            return currentInstance?.bridgeId
        }
    }

    /**
     * Reset the instance (mainly for testing)
     */
    fun resetInstance() {
        synchronized(lock) {
            currentInstance = null
        }
    }
}
